104.md

```diff
@@ -0,0 +1,226 @@
+## Group Events
+
+Group Events are all the messages that are sent within a group. This includes all "control" events that update the shared group state over time (`Proposal`, `Commit`) and messages sent between members of the group (`Application` messages).
+
+Group Events are published using an ephemeral Nostr keypair to obfuscate the number and identity of group participants. Clients MUST use a new Nostr keypair for each Group Event they publish.
+
+```json
+{
+   "id": <id>,
+   "kind": 445,
+   "created_at": <unix timestamp in seconds>,
+   "pubkey": <ephemeral sender pubkey>,
+   "content": <serialized MLSMessage object>,
+   "tags": [
+      ["h", <group id>]
+   ],
+   "sig": <signed with ephemeral sender key>
+}
+```
+- The `content` field is a [tls-style](https://www.rfc-editor.org/rfc/rfc9420.html#name-the-message-mls-media-type) serialized [`MLSMessage`](https://www.rfc-editor.org/rfc/rfc9420.html#section-6-4) object.
+- The `pubkey` is the hex-encoded public key of the ephemeral sender.
+- The `h` tag is the group ID value
+
+### Application Messages
+
+Application messages are the messages that are sent within the group by members. These are contained within the `MLSMessage` object. The format of these messages should be unsigned Nostr events of the appropriate kind. For example, if a user sends a text note to the group, it would be a `kind: 1` event. If the user reacts to a message, it would be a `kind: 7` event.
+
+This means that once the application message has been deserialized, clients can store those events and treat them as any other Nostr event, effectively creating a private Nostr feed of the group's activity and taking advantage of all the features of Nostr.
+
+The Nostr event MUST use the member's Nostr identity key for the `pubkey` field.
+
+These Nostr events MUST remain **unsigned** to ensure that if they were to leak to relays they would not be published publicly.
+
+### `Commit` Message race conditions
+
+The MLS protocol is resilient to almost all messages arriving out of order. However, the order of `Commit` messages is important for the group state to move forward from one epoch to the next correctly. Given Nostr's nature as a decentralized network, it is possible for a client to receive 2 or more `Commit` messages all attempting to update to a new epoch at the same time.
+
+Clients sending commit messages MUST wait until they receive acknowledgement from at least one relay that their Group Message Event with the `Commit` has been received before applying the commit to their own group state.
+
+If a client receives 2 or more `Commit` messages attempting to change same epoch, they MUST apply only one of the `Commit` messages they receive, determined by the following:
+
+1. Using the `created_at` timestamp on the kind `445` event. The `Commit` with the lowest value for `created_at` is the message to be applied. The other `Commit` message is discarded.
```

Comment:
Malicious clients can chose to send lower created_at values on purpose, making sure their commits are the ones that are used. 
 Maybe we can use NIP-03 (timestamp attestation) to make sure these timestamps are valid?

Reply:
I think that the time it takes to get the commit OTS attested will cause more troubles than it solves. 

Fundamentally, a malicious client that is part of the group was added to the group by someone in the group. Anyone in the group can evict that malicious client as soon as they notice bad behavior. I think this is probably not something the spec can solve and rather is the domain of the applications that are implementing this spec to decide.

Reply:
Is there someway to have an "authority" identified for these Commits? like a group admin or similar.

Reply:
> I think that the time it takes to get the commit OTS attested will cause more troubles than it solves.
> 
> Fundamentally, a malicious client that is part of the group was added to the group by someone in the group. Anyone in the group can evict that malicious client as soon as they notice bad behavior. I think this is probably not something the spec can solve and rather is the domain of the applications that are implementing this spec to decide.

Could a malicious client use a similar technique to preempt attempts to remove it?